<div id="tags">bioinformatics, biopython</div>

<h1 id="python-blast-tutorial">Python BLAST Tutorial</h1>
<h2 id="overview">Overview</h2>
<p>In this tutorial, you learn how to run BLAST <strong>locally, many times</strong> and to <strong>read BLAST results with Python</strong>. While doing so, you will familiarize with the concept of <strong>program pipelines</strong> and how you can implement them in Python.</p>
<p>The tutorial consists of six parts:</p>
<ol style="list-style-type: decimal">
<li>Preparations</li>
<li>Running local BLAST manually</li>
<li>Running local BLAST from Python</li>
<li>Running BLAST many times with Python</li>
<li>Reading BLAST output with Biopython</li>
<li>Plotting the results</li>
</ol>
<hr />
<h2 id="case-study-plasmodium-falciparum">Case Study: Plasmodium falciparum</h2>
<p>We have the hypothesis that <em>Plasmodium falciparum</em> has adapted to the human organism during its long history as a parasite. Specifically, we want to examine whether proteins from <em>Plasmodium</em> are more similar to human proteins than one would expect. If this is true we could interpret e.g. that more similar proteins help <em>Plasmodium</em> to evade the human immune system.</p>
<p>As a small sample study, we will BLAST a set of peptides from a few <em>Homo sapiens</em> proteins against the proteome of <em>Plasmodium falciparum</em>. As a control, we will use the proteome of <em>Schizosaccharomyces pombe</em>.</p>
<hr />
<h2 id="preparations">1. Preparations</h2>
<h3 id="check-whether-blast-is-properly-installed">1.1 Check whether BLAST+ is properly installed</h3>
<p>Enter the two following commands in a Linux console:</p>
<pre><code>makeblastdb
blastp</code></pre>
<p>Both should result in an error message other than <em>command not found</em>.</p>
<h3 id="create-a-blast-database-for-plasmodium-falciparum">1.2 Create a BLAST database for Plasmodium falciparum</h3>
<p>Create a BLAST database for the <em>Plasmodium</em> proteins. First, open a console and go to the folder <em>data/</em> . Type:</p>
<pre><code>makeblastdb -in Plasmodium_falciparum.fasta -dbtype prot</code></pre>
<p>You should see a message similar to:</p>
<pre><code>Adding sequences from FASTA; added 5414 sequences in 0.56993 seconds.</code></pre>
<h3 id="create-a-blast-database-for-the-control-organism">1.3 Create a BLAST database for the control organism</h3>
<p>Please create a BLAST database for <em>Schizosaccharomyces pombe</em> as well.</p>
<h3 id="questions">1.4 Questions</h3>
<ul>
<li>What files have appeared in the <em>data/</em> directory?</li>
<li>Why do we need to create a database first? Why can't BLAST do that right before each query?</li>
</ul>
<hr />
<h2 id="running-local-blast-manually">2. Running local BLAST manually</h2>
<p>Before running a large series of BLAST experiments, we will run a small sequence as a technical proof of concept. We are using a sequence copied from the <em>Plasmodium</em> sequences, so we know that BLAST should generate a 100% match.</p>
<h3 id="create-a-query-file">2.1 Create a query file</h3>
<p>Create an empty file <em>query.seq</em> in a text editor. Write the following peptide sequence into the file:</p>
<pre><code>DAAITAALNANAVK</code></pre>
<p>Make sure that there are no other characters in the file (no empty lines or FASTA deflines). Save the file to the <em>data/</em> directory.</p>
<h3 id="running-local-blast-against-plasmodium">2.2 Running local BLAST against Plasmodium</h3>
<p>Go to a console in the <em>data/</em> directory and type:</p>
<pre><code>blastp -query query.seq -db Plasmodium_falciparum.fasta -out output.txt -outfmt 7</code></pre>
<h3 id="running-local-blast-against-the-control-group">2.3 Running local BLAST against the control group</h3>
<p>Repeat the above query for <em>Schizosaccharomyces pombe</em>.</p>
<h3 id="adjust-output-formats">2.4 Adjust output formats</h3>
<p>Insert different numbers (1-7) for the <strong>outfmt</strong> parameter and re-run the query.</p>
<h3 id="questions-1">2.5 Questions</h3>
<ul>
<li>Take a look at the BLAST output. Is the result what you would expect?</li>
<li>Does the control group support your assumptions so far?</li>
<li>Which of the output formats do you find the easiest to read?</li>
<li>Which of the output formats is probably the easiest to read for a program?</li>
</ul>
<hr />
<h2 id="running-local-blast-from-python">3. Running local BLAST from Python</h2>
<p>Now we are going to do exactly the same operation from a Python program. For this we will need the <strong>os module</strong>.</p>
<h3 id="introduction-to-the-os-module">3.1 Introduction to the os module</h3>
<p>Open the document <em>pipelines/os_module_puzzle.pdf</em>. Do the exercise.</p>
<h3 id="running-blast-from-python">3.2 Running BLAST from Python</h3>
<p>Now we will use the function <strong>os.system</strong> to run BLAST. Create a Python script <strong>run_blast.py</strong> in the <em>data/</em> directory. Write the following commands into it:</p>
<pre><code>import os

cmd = &quot;blastp -query query.seq -db Plasmodium_falciparum.fasta -out output.txt -outfmt 7&quot;
os.system(cmd)</code></pre>
<p>Execute the program.</p>
<h3 id="customizing-the-query">3.3 Customizing the query</h3>
<p>In order to make the BLAST command in Python more flexible, we will combine it from variables. Change the code to the following:</p>
<pre><code>db = &quot;Plasmodium_falciparum.fasta&quot;
cmd = &quot;blastp -query query.seq -db &quot; + db + &quot; -out output.txt -outfmt 7&quot;</code></pre>
<h3 id="more-variables">3.4 More variables</h3>
<p>Now add separate variables for the query and the output file name as well.</p>
<h3 id="additional-examples-for-using-os">3.5 Additional examples for using os</h3>
<p>In the <em>pipelines/</em> directory you find more examples using the <em>os module</em>. If you like, try them out as well.</p>
<h3 id="questions-2">3.6 Questions</h3>
<ul>
<li>Is the output of the Python BLAST run identical to the one you did manually? How can you check that?</li>
</ul>
<hr />
<h2 id="running-blast-many-times-with-python">4. Running BLAST many times with Python</h2>
<h3 id="creating-query-files">4.1 Creating query files</h3>
<p>The file <em>data/human_peptide.fasta</em> contains about 2000 peptides. We want to run BLAST for each of them. To do so, we need to write each peptide to a separate file.</p>
<p>First, create a new folder for the query files:</p>
<pre><code>mkdir data/queries</code></pre>
<p>The Python script <strong>multiblast/split_fasta.py</strong> does that using <strong>Bio.SeqIO</strong>. You can use it by typing in the <em>multiblast/</em> directory:</p>
<pre><code>python split_fasta.py ../data/human_peptide.fasta ../data/queries</code></pre>
<p>If you want, you can try writing that script by yourself.</p>
<h3 id="validate-the-queries">4.2 Validate the queries</h3>
<p>Make sure that the query files have been generated and that they are not empty. You can check both with:</p>
<pre><code>ls -l data/queries
more data/queries/9568103_99.fasta</code></pre>
<h3 id="create-output-directories">4.3 Create output directories</h3>
<p>Prepare a place where the results from each BLAST run will be stored:</p>
<pre><code>mkdir data/Plasmodium_out
mkdir data/Pombe_out</code></pre>
<h3 id="run-blast">4.4 Run BLAST</h3>
<p>You can run BLAST for all queries with the script <strong>multiblast/run_blast.py</strong>. It uses <strong>os</strong> for three different things:</p>
<ol style="list-style-type: decimal">
<li>Reading directory names as command-line parameters</li>
<li>Looping through all files in a directory</li>
<li>Running the BLAST command</li>
</ol>
<p><strong>However, the program is incomplete.</strong></p>
<p>You need to complete the BLAST command inserting the file names from the given variables. Use the parameter <em>-outfmt 5</em> in order to create XML output. We will need this later to read it from Biopython.</p>
<p>When everything is done, you should be able to execute the script with:</p>
<pre><code>python run_blast.py ../data/queries/ ../data/Plasmodium_falciparum.fasta ../data/Plasmodium_out/</code></pre>
<p>Inspect the result.</p>
<hr />
<h2 id="reading-blast-output-with-biopython">5. Reading BLAST output with Biopython</h2>
<h3 id="reading-xml-data">5.1 Reading XML data</h3>
<p>Run the program <strong>BLAST_XML/parse_blast_xml.py</strong>.</p>
<pre><code>python parse_blast_xml.py</code></pre>
<h3 id="read-one-of-your-blast-result-files">5.2 Read one of your BLAST result files</h3>
<p>Adjust the program to read one of your BLAST output files. Try to figure out how many HSPs there are, and how many are below an e-value threshold of 0.001.</p>
<h3 id="read-all-of-your-blast-result-files">5.3 Read all of your BLAST result files</h3>
<p>Customize the program to read <strong>all</strong> of your result files. How many hits do you have in total. What is the hit with the highest score?</p>
<hr />
<h2 id="references">References</h2>
<p>BLAST+ is a new, faster (C++ based) version that replaces BLAST2, as of Oct 2013. Also see: http://blast.ncbi.nlm.nih.gov/Blast.cgi?CMD=Web&amp;PAGE_TYPE=BlastDocs&amp;DOC_TYPE=Download</p>
